# Data preprocessing

I will work with 5 dataframes (2015-2019). 

In order to be able to perform comparative analysis between different years I need to preprocess my data first. 

I will be checkign for the null values and making sure that I have same columns across dataframes. 

For the purpose of comparative analyis I will keep only those  countries that are in common for all 5 dataframes. I assume that there will not be many of those that are different as I do not want to drop too many data.I will adjust my dataframes and saved them under new names. In these new dataframes I will also add the region.


In [1]:
#importing libraries for data preprocessing
import pandas as pd

In [2]:
#loading datasets

df1=pd.read_csv("/Users/Jordanka/Documents/PD/My portfolios/Polish those/Happiness/2015.csv")
df2=pd.read_csv("/Users/Jordanka/Documents/PD/My portfolios/Polish those/Happiness/2016.csv")
df3=pd.read_csv("/Users/Jordanka/Documents/PD/My portfolios/Polish those/Happiness/2017.csv")
df4=pd.read_csv("/Users/Jordanka/Documents/PD/My portfolios/Polish those/Happiness/2018.csv")
df5=pd.read_csv("/Users/Jordanka/Documents/PD/My portfolios/Polish those/Happiness/2019.csv")

In [3]:
#chechking for columns
print(df1.columns,'\n',df2.columns,'\n',df3.columns,'\n',df4.columns,'\n',df5.columns)

Index(['Country', 'Region', 'Happiness Rank', 'Happiness Score',
       'Standard Error', 'Economy (GDP per Capita)', 'Family',
       'Health (Life Expectancy)', 'Freedom', 'Trust (Government Corruption)',
       'Generosity', 'Dystopia Residual'],
      dtype='object') 
 Index(['Country', 'Region', 'Happiness Rank', 'Happiness Score',
       'Lower Confidence Interval', 'Upper Confidence Interval',
       'Economy (GDP per Capita)', 'Family', 'Health (Life Expectancy)',
       'Freedom', 'Trust (Government Corruption)', 'Generosity',
       'Dystopia Residual'],
      dtype='object') 
 Index(['Country', 'Happiness.Rank', 'Happiness.Score', 'Whisker.high',
       'Whisker.low', 'Economy..GDP.per.Capita.', 'Family',
       'Health..Life.Expectancy.', 'Freedom', 'Generosity',
       'Trust..Government.Corruption.', 'Dystopia.Residual'],
      dtype='object') 
 Index(['Overall rank', 'Country or region', 'Score', 'GDP per capita',
       'Social support', 'Healthy life expectancy',
     

In [4]:
df1.head(5)

Country          Region  Happiness Rank  Happiness Score  \
0  Switzerland  Western Europe               1            7.587   
1      Iceland  Western Europe               2            7.561   
2      Denmark  Western Europe               3            7.527   
3       Norway  Western Europe               4            7.522   
4       Canada   North America               5            7.427   

   Standard Error  Economy (GDP per Capita)   Family  \
0         0.03411                   1.39651  1.34951   
1         0.04884                   1.30232  1.40223   
2         0.03328                   1.32548  1.36058   
3         0.03880                   1.45900  1.33095   
4         0.03553                   1.32629  1.32261   

   Health (Life Expectancy)  Freedom  Trust (Government Corruption)  \
0                   0.94143  0.66557                        0.41978   
1                   0.94784  0.62877                        0.14145   
2                   0.87464  0.64938                        0.48357   
3                   0.88521  0.66973                        0.36503   
4                   0.90563  0.63297                        0.32957   

   Generosity  Dystopia Residual  
0     0.29678            2.51738  
1     0.43630            2.70201  
2     0.34139            2.49204  
3     0.34699            2.46531  
4     0.45811            2.45176

In [5]:
df2.head(5)

Country          Region  Happiness Rank  Happiness Score  \
0      Denmark  Western Europe               1            7.526   
1  Switzerland  Western Europe               2            7.509   
2      Iceland  Western Europe               3            7.501   
3       Norway  Western Europe               4            7.498   
4      Finland  Western Europe               5            7.413   

   Lower Confidence Interval  Upper Confidence Interval  \
0                      7.460                      7.592   
1                      7.428                      7.590   
2                      7.333                      7.669   
3                      7.421                      7.575   
4                      7.351                      7.475   

   Economy (GDP per Capita)   Family  Health (Life Expectancy)  Freedom  \
0                   1.44178  1.16374                   0.79504  0.57941   
1                   1.52733  1.14524                   0.86303  0.58557   
2                   1.42666  1.18326                   0.86733  0.56624   
3                   1.57744  1.12690                   0.79579  0.59609   
4                   1.40598  1.13464                   0.81091  0.57104   

   Trust (Government Corruption)  Generosity  Dystopia Residual  
0                        0.44453     0.36171            2.73939  
1                        0.41203     0.28083            2.69463  
2                        0.14975     0.47678            2.83137  
3                        0.35776     0.37895            2.66465  
4                        0.41004     0.25492            2.82596

In [6]:
# checking if the confidence interval is indeed mean +- st.error

(7526-7460)/1.96==(7592-7526)/1.96

True

In [7]:
#it is so we can actually get the st.error from these 2 columns if needed

df3.head(5)

Country  Happiness.Rank  Happiness.Score  Whisker.high  Whisker.low  \
0       Norway               1            7.537      7.594445     7.479556   
1      Denmark               2            7.522      7.581728     7.462272   
2      Iceland               3            7.504      7.622030     7.385970   
3  Switzerland               4            7.494      7.561772     7.426227   
4      Finland               5            7.469      7.527542     7.410458   

   Economy..GDP.per.Capita.    Family  Health..Life.Expectancy.   Freedom  \
0                  1.616463  1.533524                  0.796667  0.635423   
1                  1.482383  1.551122                  0.792566  0.626007   
2                  1.480633  1.610574                  0.833552  0.627163   
3                  1.564980  1.516912                  0.858131  0.620071   
4                  1.443572  1.540247                  0.809158  0.617951   

   Generosity  Trust..Government.Corruption.  Dystopia.Residual  
0    0.362012                       0.315964           2.277027  
1    0.355280                       0.400770           2.313707  
2    0.475540                       0.153527           2.322715  
3    0.290549                       0.367007           2.276716  
4    0.245483                       0.382612           2.430182

df3 doesn't have confidence interval but rather IQR , so  td.error can be dropped from further analysis

In [8]:
df4.head(5)

Overall rank Country or region  Score  GDP per capita  Social support  \
0             1           Finland  7.632           1.305           1.592   
1             2            Norway  7.594           1.456           1.582   
2             3           Denmark  7.555           1.351           1.590   
3             4           Iceland  7.495           1.343           1.644   
4             5       Switzerland  7.487           1.420           1.549   

   Healthy life expectancy  Freedom to make life choices  Generosity  \
0                    0.874                         0.681       0.202   
1                    0.861                         0.686       0.286   
2                    0.868                         0.683       0.284   
3                    0.914                         0.677       0.353   
4                    0.927                         0.660       0.256   

   Perceptions of corruption  
0                      0.393  
1                      0.340  
2                      0.408  
3                      0.138  
4                      0.357

In [9]:
df5.head()

Overall rank Country or region  Score  GDP per capita  Social support  \
0             1           Finland  7.769           1.340           1.587   
1             2           Denmark  7.600           1.383           1.573   
2             3            Norway  7.554           1.488           1.582   
3             4           Iceland  7.494           1.380           1.624   
4             5       Netherlands  7.488           1.396           1.522   

   Healthy life expectancy  Freedom to make life choices  Generosity  \
0                    0.986                         0.596       0.153   
1                    0.996                         0.592       0.252   
2                    1.028                         0.603       0.271   
3                    1.026                         0.591       0.354   
4                    0.999                         0.557       0.322   

   Perceptions of corruption  
0                      0.393  
1                      0.410  
2                      0.341  
3                      0.118  
4                      0.298

df4 and df5 are missing distopia residual. when I check through sample calculation it seems that we can find it by subtracting all other values from the Happiness score like in the example below fro Finland in df5 :

7.769-1.340-1.587-0.986-0.596-0.153-0.393= 2.714

This seems to be a reasonable value when Icompare it with the residuals from the first 3 data frames. I will therefor make a residual columns for df4 and df5.

In [10]:
df4['residual']=df4['Score'] - df4['GDP per capita']-df4['Social support'] - df4['Healthy life expectancy']- df4['Freedom to make life choices'] - df4['Generosity']-df4['Perceptions of corruption']

In [11]:
df5['residual']=df5['Score'] - df5['GDP per capita']-df5['Social support'] - df5['Healthy life expectancy']- df5['Freedom to make life choices'] - df5['Generosity']-df5['Perceptions of corruption']

In [12]:
df4.head()

Overall rank Country or region  Score  GDP per capita  Social support  \
0             1           Finland  7.632           1.305           1.592   
1             2            Norway  7.594           1.456           1.582   
2             3           Denmark  7.555           1.351           1.590   
3             4           Iceland  7.495           1.343           1.644   
4             5       Switzerland  7.487           1.420           1.549   

   Healthy life expectancy  Freedom to make life choices  Generosity  \
0                    0.874                         0.681       0.202   
1                    0.861                         0.686       0.286   
2                    0.868                         0.683       0.284   
3                    0.914                         0.677       0.353   
4                    0.927                         0.660       0.256   

   Perceptions of corruption  residual  
0                      0.393     2.585  
1                      0.340     2.383  
2                      0.408     2.371  
3                      0.138     2.426  
4                      0.357     2.318

In [13]:
df5.head()

Overall rank Country or region  Score  GDP per capita  Social support  \
0             1           Finland  7.769           1.340           1.587   
1             2           Denmark  7.600           1.383           1.573   
2             3            Norway  7.554           1.488           1.582   
3             4           Iceland  7.494           1.380           1.624   
4             5       Netherlands  7.488           1.396           1.522   

   Healthy life expectancy  Freedom to make life choices  Generosity  \
0                    0.986                         0.596       0.153   
1                    0.996                         0.592       0.252   
2                    1.028                         0.603       0.271   
3                    1.026                         0.591       0.354   
4                    0.999                         0.557       0.322   

   Perceptions of corruption  residual  
0                      0.393     2.714  
1                      0.410     2.394  
2                      0.341     2.241  
3                      0.118     2.401  
4                      0.298     2.394

I will cpompare the average residual for df4 and df5 with residuals from previous years to check if the values seems reasonable.

In [14]:
print('' ,df1['Dystopia Residual'].mean(),'\n',df2['Dystopia Residual'].mean(),'\n',df3['Dystopia.Residual'].mean(),
      '\n', df4['residual'].mean(), '\n',df5['residual'].mean())

 2.098976772151899 
 2.3258065605095544 
 1.8502378056126256 
 1.9217032258064526 
 1.8798717948717951


Values do indeed seem reasonable. 

We can notice that the column family has been replaced with the Social support but the values are very similar. I believe that it is safe to assume that those columns represent the same thing as sometimes people consider very close friend the family and probably think of them as of family. So I will treat these 2 columns as they're representing the same thing.

There are 8  common columns amongst these 5 dataframes:
-country or region,score,GDP,Family/social support,life expectancy,freedom to make choices,generosity and perception of corruption.There is no region column in the last 3 dataframes and I think that it is an important coulumn my future analysis. I will add the region column after I clean my datasets. 

I will rename columns and drop columns that are not in common (except from region). I will save those dataframes as df1_new,df2_new... 

In [15]:
columns=['Country', 'Region','Happiness Score','Economy (GDP per Capita)', 'Family',
       'Health (Life Expectancy)', 'Freedom', 'Trust (Government Corruption)','Generosity','Dystopia Residual']

In [16]:
new_columns_15=['country','region','happiness score','GDP_15','family/social_support_15','life_expectancy_15','freedom_to_make_choices_15','perception_of_corruption_15','generosity_15','residual_15']

In [17]:
df1_new=df1.drop(['Happiness Rank','Standard Error'],axis=1)

In [18]:
df1_new.rename(columns=dict(zip(columns,new_columns_15)),inplace=True)

In [19]:
df1_new

country                           region  happiness score   GDP_15  \
0    Switzerland                   Western Europe            7.587  1.39651   
1        Iceland                   Western Europe            7.561  1.30232   
2        Denmark                   Western Europe            7.527  1.32548   
3         Norway                   Western Europe            7.522  1.45900   
4         Canada                    North America            7.427  1.32629   
..           ...                              ...              ...      ...   
153       Rwanda               Sub-Saharan Africa            3.465  0.22208   
154        Benin               Sub-Saharan Africa            3.340  0.28665   
155        Syria  Middle East and Northern Africa            3.006  0.66320   
156      Burundi               Sub-Saharan Africa            2.905  0.01530   
157         Togo               Sub-Saharan Africa            2.839  0.20868   

     family/social_support_15  life_expectancy_15  freedom_to_make_choices_15  \
0                     1.34951             0.94143                     0.66557   
1                     1.40223             0.94784                     0.62877   
2                     1.36058             0.87464                     0.64938   
3                     1.33095             0.88521                     0.66973   
4                     1.32261             0.90563                     0.63297   
..                        ...                 ...                         ...   
153                   0.77370             0.42864                     0.59201   
154                   0.35386             0.31910                     0.48450   
155                   0.47489             0.72193                     0.15684   
156                   0.41587             0.22396                     0.11850   
157                   0.13995             0.28443                     0.36453   

     perception_of_corruption_15  generosity_15  residual_15  
0                        0.41978        0.29678      2.51738  
1                        0.14145        0.43630      2.70201  
2                        0.48357        0.34139      2.49204  
3                        0.36503        0.34699      2.46531  
4                        0.32957        0.45811      2.45176  
..                           ...            ...          ...  
153                      0.55191        0.22628      0.67042  
154                      0.08010        0.18260      1.63328  
155                      0.18906        0.47179      0.32858  
156                      0.10062        0.19727      1.83302  
157                      0.10731        0.16681      1.56726  

[158 rows x 10 columns]

In [20]:
columns_2=['Country','Region', 'Happiness Score','Economy (GDP per Capita)', 'Family', 'Health (Life Expectancy)',
       'Freedom', 'Trust (Government Corruption)', 'Generosity',
       'Dystopia Residual']
df2_new=df2.drop(['Happiness Rank','Lower Confidence Interval', 'Upper Confidence Interval'],axis=1)

In [21]:
new_columns_16=['country','region','happiness score','GDP_16','family/social_support_16','life_expectancy_16','freedom_to_make_choices_16','perception_of_corruption_16','generosity_16','residual_16']

In [22]:
df2_new.rename(columns=dict(zip(columns_2,new_columns_16)),inplace=True)

In [23]:
df2_new.head(2)

country          region  happiness score   GDP_16  \
0      Denmark  Western Europe            7.526  1.44178   
1  Switzerland  Western Europe            7.509  1.52733   

   family/social_support_16  life_expectancy_16  freedom_to_make_choices_16  \
0                   1.16374             0.79504                     0.57941   
1                   1.14524             0.86303                     0.58557   

   perception_of_corruption_16  generosity_16  residual_16  
0                      0.44453        0.36171      2.73939  
1                      0.41203        0.28083      2.69463

In [24]:
columns_3=['Country', 'Happiness.Score', 'Economy..GDP.per.Capita.', 'Family',
       'Health..Life.Expectancy.', 'Freedom','Trust..Government.Corruption.',
            'Generosity','Dystopia.Residual']
df3_new=df3.drop([  'Happiness.Rank','Whisker.high','Whisker.low', ],axis=1)
new_columns_17=['country','happiness score','GDP_17','family/social_support_17','life_expectancy_17','freedom_to_make_choices_17','perception_of_corruption_17','generosity_17','residual_17']
df3_new.rename(columns=dict(zip(columns_3,new_columns_17)),inplace=True)

In [25]:
df3_new.head()

country  happiness score    GDP_17  family/social_support_17  \
0       Norway            7.537  1.616463                  1.533524   
1      Denmark            7.522  1.482383                  1.551122   
2      Iceland            7.504  1.480633                  1.610574   
3  Switzerland            7.494  1.564980                  1.516912   
4      Finland            7.469  1.443572                  1.540247   

   life_expectancy_17  freedom_to_make_choices_17  generosity_17  \
0            0.796667                    0.635423       0.362012   
1            0.792566                    0.626007       0.355280   
2            0.833552                    0.627163       0.475540   
3            0.858131                    0.620071       0.290549   
4            0.809158                    0.617951       0.245483   

   perception_of_corruption_17  residual_17  
0                     0.315964     2.277027  
1                     0.400770     2.313707  
2                     0.153527     2.322715  
3                     0.367007     2.276716  
4                     0.382612     2.430182

In [26]:
#makin order of columns the same
#df3_new=df3_new[new_columns_17]

In [27]:
#df3_new.head(2)

In [28]:
#I will rename and drop columns for df4 and df5
columns_4=['Country or region', 'Score', 'GDP per capita',
       'Social support', 'Healthy life expectancy',
       'Freedom to make life choices', 'Generosity',
       'Perceptions of corruption','residual']
df4_new=df4.drop(['Overall rank'],axis=1)
new_columns_18=['country','happiness score','GDP_18','family/social_support_18','life_expectancy_18','freedom_to_make_choices_18','perception_of_corruption_18','generosity_18','residual_18']
df4_new.rename(columns=dict(zip(columns_4,new_columns_18)),inplace=True)

In [29]:
df4_new.head()

country  happiness score  GDP_18  family/social_support_18  \
0      Finland            7.632   1.305                     1.592   
1       Norway            7.594   1.456                     1.582   
2      Denmark            7.555   1.351                     1.590   
3      Iceland            7.495   1.343                     1.644   
4  Switzerland            7.487   1.420                     1.549   

   life_expectancy_18  freedom_to_make_choices_18  \
0               0.874                       0.681   
1               0.861                       0.686   
2               0.868                       0.683   
3               0.914                       0.677   
4               0.927                       0.660   

   perception_of_corruption_18  generosity_18  residual_18  
0                        0.202          0.393        2.585  
1                        0.286          0.340        2.383  
2                        0.284          0.408        2.371  
3                        0.353          0.138        2.426  
4                        0.256          0.357        2.318

In [30]:
columns_5=['Country or region', 'Score', 'GDP per capita',
       'Social support', 'Healthy life expectancy',
       'Freedom to make life choices', 'Generosity',
       'Perceptions of corruption','residual']
df5_new=df5.drop([  'Overall rank' ],axis=1)
new_columns_19=['country','happiness score','GDP_19','family/social_support_19','life_expectancy_19','freedom_to_make_choices_19','perception_of_corruption_19','generosity_19','residual_19']
df5_new.rename(columns=dict(zip(columns_5,new_columns_19)),inplace=True)



In [31]:
df5_new.head(2)

country  happiness score  GDP_19  family/social_support_19  \
0  Finland            7.769   1.340                     1.587   
1  Denmark            7.600   1.383                     1.573   

   life_expectancy_19  freedom_to_make_choices_19  \
0               0.986                       0.596   
1               0.996                       0.592   

   perception_of_corruption_19  generosity_19  residual_19  
0                        0.153          0.393        2.714  
1                        0.252          0.410        2.394

Now we have same columns. Let's check for null data and data types.

In [32]:
print(df1_new.info(),'\n',df2_new.info(),'\\n',df3_new.info(),
      '\\n',df4_new.info(),'\\n',df5_new.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 158 entries, 0 to 157
Data columns (total 10 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   country                      158 non-null    object 
 1   region                       158 non-null    object 
 2   happiness score              158 non-null    float64
 3   GDP_15                       158 non-null    float64
 4   family/social_support_15     158 non-null    float64
 5   life_expectancy_15           158 non-null    float64
 6   freedom_to_make_choices_15   158 non-null    float64
 7   perception_of_corruption_15  158 non-null    float64
 8   generosity_15                158 non-null    float64
 9   residual_15                  158 non-null    float64
dtypes: float64(8), object(2)
memory usage: 12.5+ KB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 157 entries, 0 to 156
Data columns (total 10 columns):
 #   Column                       Non-Nul

There are no null data and all numbers are floats. 

But the df3_new has 155 countires and the df1_new has
158 countries. The next step is to reduce the number of countries 
to the highest common number of countries for all datasets. I will start by comparing df3_new( because it has the smallest number of data )with the  df1_new( largest number of data.)

In [33]:
n=0
for country in df1_new['country'].unique() :
    if country not in df3_new['country'].unique():
        n+=1
        print( country,False)
print(f'\n{n} countries are not in df3_new')

Oman False
Taiwan False
Suriname False
Hong Kong False
Somaliland region False
Laos False
Swaziland False
Djibouti False
Comoros False

9 countries are not in df3_new


In [34]:
#now the wice versa
n=0
for country in df3_new['country'].unique() :
    if country not in df1_new['country'].unique():
        n+=1
        print( country,False)
print(f'\n{n} countries are not in df1_new')

Taiwan Province of China False
Belize False
Hong Kong S.A.R., China False
Somalia False
Namibia False
South Sudan False

6 countries are not in df1_new


From here I can see that some countries have similar names. I will drop only those countries that are actusally different countries and I will rename ones that have simlar names.

In [35]:
#dropping countries that are in df3_new but not in the df1_new
to_drop=['Oman','Suriname','Laos','Swaziland','Djibouti','Comoros','Somaliland region']
for name in to_drop:
    df1_new.drop( df1_new[ df1_new['country'] == name ].index , inplace=True)

In [36]:
len(df1_new)

151

In [37]:
#dropping countries that are in df1_new but not in the df3_new
to_drop1=['Belize','Namibia','South Sudan','Somalia']
for name in to_drop1:
    df3_new.drop( df3_new[ df3_new['country'] == name ].index , inplace=True)

In [38]:
len(df3_new)

151

In [39]:
df3_new=df3_new.replace(['Taiwan Province of China','Hong Kong S.A.R., China'],['Taiwan','Hong Kong'])

I will now compare df1_comp with the rest of df and I hope that I will not have to delete many more  coumns from df1_comp

In [40]:
n=0
for country in df1_new['country'].unique() :
    if country not in df2_new['country'].unique():
        n+=1
        print( country,False)
print(f'\n{n} countries are not in df2_new')

Mozambique False
Lesotho False
Central African Republic False

3 countries are not in df2_new


In [41]:
n=0
for country in df2_new['country'].unique() :
    if country not in df1_new['country'].unique():
        n+=1
        print( country,False)
print(f'\n{n} countries are not in df1_new')

Puerto Rico False
Suriname False
Belize False
Somalia False
Somaliland Region False
Laos False
Namibia False
Comoros False
South Sudan False

9 countries are not in df1_new


From above I can see that Iwill need to delete 3 more countries from df1_comp and df3_comp and 9 countries from df2_comp

In [42]:
#dropping countries that are in df2_new but not in the df1_new (and therefore not in df3_new)
to_drop2=['Mozambique','Lesotho','Central African Republic' ]
for name in to_drop2:
    df1_new.drop( df1_new[ df1_new['country'] == name ].index , inplace=True)
    df3_new.drop( df3_new[ df3_new['country'] == name ].index , inplace=True)

In [43]:
#dropping countries that are in df1_new but not in the df2_new 
to_drop3=['Puerto Rico', 'Suriname', 'Belize','Somalia', 'Laos','Namibia', 'Comoros', 'South Sudan','Somaliland Region']
for name in to_drop3:
    df2_new.drop( df2_new[ df2_new['country'] == name ].index , inplace=True)

In [44]:
print(len(df1_new),len(df2_new),len(df3_new))

148 148 148


In [45]:
n=0
for country in df1_new['country'].unique() :
    if country not in df4_new['country'].unique():
        n+=1
        print( country,False)
print(f'\n{n} countries are not in df4_new')

Trinidad and Tobago False
North Cyprus False

2 countries are not in df4_new


In [46]:
n=0
for country in df4_new['country'].unique() :
    if country not in df1_new['country'].unique():
        n+=1
        print( country,False)
print(f'\n{n} countries are not in df1_new')

Trinidad & Tobago False
Belize False
Northern Cyprus False
Somalia False
Laos False
Namibia False
Mozambique False
Lesotho False
South Sudan False
Central African Republic False

10 countries are not in df1_new


So we need to rename 2 countries in df4_comp as Trinidad and Tobago and NOrth Cyprus have been used in df1, df2 and df3 . We also need to delete 8 countries from df4_comp

In [47]:
#renaming
df4_new=df4_new.replace(['Trinidad & Tobago','Northern Cyprus'],
                  ['Trinidad and Tobago','North Cyprus'])

In [48]:
#dropping countries that are in df4_new but not in the df1_new
to_drop4=['Belize','Somalia','Laos', 'Namibia', 'South Sudan','Central African Republic',
          'Mozambique','Lesotho']
for name in to_drop4:
    df4_new.drop( df4_new[ df4_new['country'] == name ].index , inplace=True)

In [49]:
len(df4_new)

148

In [50]:
n=0
for country in df1_new['country'].unique() :
    if country not in df5_new['country'].unique():
        n+=1
        print( country,False)
print(f'\n{n} countries are not in df5_new')

Trinidad and Tobago False
North Cyprus False
Macedonia False
Sudan False
Angola False

5 countries are not in df5_new


In [51]:
n=0
for country in df5_new['country'].unique() :
    if country not in df1_new['country'].unique():
        n+=1
        print( country,False)
print(f'\n{n} countries are not in df1_new')

Trinidad & Tobago False
Northern Cyprus False
North Macedonia False
Laos False
Somalia False
Namibia False
Gambia False
Mozambique False
Swaziland False
Comoros False
Lesotho False
Central African Republic False
South Sudan False

13 countries are not in df1_new


In [52]:
#renaming countries
df5_new=df5_new.replace(['Trinidad & Tobago','Northern Cyprus','North Macedonia'],
                  ['Trinidad and Tobago','North Cyprus','Macedonia'])

In [53]:
#dropping countries that are in df1_new but not in the df5_new
to_drop5=['Laos','Somalia','Namibia','Gambia','Swaziland','Comoros', 'Central African Republic','South Sudan','Mozambique','Lesotho']
for name in to_drop5:
    df5_new.drop( df5_new[ df5_new['country'] == name ].index , inplace=True)

In [54]:
#droppiing countries that are in df5_nwe but not in the other datasets
to_drop6=['Sudan','Angola']
data_frames=[df1_new,df2_new,df3_new,df4_new]
for name in to_drop6:
    for frame in data_frames:
        frame.drop( frame[ frame['country'] == name ].index , inplace=True)


In [55]:
print(len(df1_new),len(df2_new),len(df3_new),len(df4_new),len(df5_new))

146 146 146 146 146


12 countries have been lost in data cleaning. It is important to mention that those countries are still in the original dataframes and they will be displayed on geomaps. But for the sake of comparison between years it is appropriate to have same countries.  
Now I will sort data by country to check once more if all dataframes have the same countries 

In [56]:
df1_new=df1_new.sort_values('country')
df2_new=df2_new.sort_values('country')
df3_new=df3_new.sort_values('country')
df4_new=df4_new.sort_values('country')
df5_new=df5_new.sort_values('country')

In [57]:
df1_new['country'].unique()==df2_new['country'].unique()

array([ True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,

In [58]:
df1_new['country'].unique()==df3_new['country'].unique()

array([ True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,

In [59]:
df1_new['country'].unique()==df4_new['country'].unique()

array([ True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,

In [60]:
df1_new['country'].unique()==df5_new['country'].unique()

array([ True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,

So, we can now merge those dataframes if needed. But first let me reset the index

In [61]:
frames=[df1_new,df2_new,df3_new,df4_new,df5_new]
for frame in frames:
    frame=frame.reset_index(drop=True,inplace=True)
    

In [62]:
df1_new.head(3)

country                           region  happiness score   GDP_15  \
0  Afghanistan                    Southern Asia            3.575  0.31982   
1      Albania       Central and Eastern Europe            4.959  0.87867   
2      Algeria  Middle East and Northern Africa            5.605  0.93929   

   family/social_support_15  life_expectancy_15  freedom_to_make_choices_15  \
0                   0.30285             0.30335                     0.23414   
1                   0.80434             0.81325                     0.35733   
2                   1.07772             0.61766                     0.28579   

   perception_of_corruption_15  generosity_15  residual_15  
0                      0.09719        0.36510      1.95210  
1                      0.06413        0.14272      1.89894  
2                      0.17383        0.07822      2.43209

I will now add region column to df3,df4 and df5. As the order of countries is the same I can simply assign the new column using the df1_new['region]. I will also change the order of columns.

In [63]:
df3_new['region']=df1_new['region']
new_col=['country','region','happiness score','GDP_17','family/social_support_17','life_expectancy_17','freedom_to_make_choices_17','perception_of_corruption_17','generosity_17','residual_17']
df3_new=df3_new[new_col]

In [64]:
df3_new.head(2)

country                      region  happiness score    GDP_17  \
0  Afghanistan               Southern Asia            3.794  0.401477   
1      Albania  Central and Eastern Europe            4.644  0.996193   

   family/social_support_17  life_expectancy_17  freedom_to_make_choices_17  \
0                  0.581543            0.180747                    0.106180   
1                  0.803685            0.731160                    0.381499   

   perception_of_corruption_17  generosity_17  residual_17  
0                     0.061158       0.311871     2.150801  
1                     0.039864       0.201313     1.490442

I will repeat the same for df4_new and df5_new.

In [65]:
df4_new['region']=df1_new['region']
new_col=['country','region','happiness score','GDP_18','family/social_support_18','life_expectancy_18','freedom_to_make_choices_18','perception_of_corruption_18','generosity_18','residual_18']
df4_new=df4_new[new_col]

In [66]:
df5_new['region']=df1_new['region']
new_col=['country','region','happiness score','GDP_19','family/social_support_19','life_expectancy_19','freedom_to_make_choices_19','perception_of_corruption_19','generosity_19','residual_19']
df5_new=df5_new[new_col]

In [67]:
%store df1_new

Stored 'df1_new' (DataFrame)


In [68]:
%store df2_new

Stored 'df2_new' (DataFrame)


In [69]:
%store df3_new

Stored 'df3_new' (DataFrame)


In [70]:
%store df4_new

Stored 'df4_new' (DataFrame)


In [71]:
%store df5_new

Stored 'df5_new' (DataFrame)
